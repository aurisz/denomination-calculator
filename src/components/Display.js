import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

import { currencyFormatter } from '../utils';

const Display = ({ result, isError }) => {
  if (isError) {
    return (
      <p>
        Invalid Input. Please try again using valid amount (e.g: 18.215,
        Rp17500, Rp 17.500,00){' '}
      </p>
    );
  }

  if (result === null) {
    return (
      <p>
        Simple React app to calculate how many rupiah notes needed to make that
        amount
      </p>
    );
  }

  return (
    <Fragment>
      {result.fractions.map((fraction, index) => (
        <div className="card" key={index}>
          <span>
            {fraction.count} x {fraction.value}
          </span>
          <span>Rp{currencyFormatter(fraction.total)}</span>
        </div>
      ))}
      {result.remaining !== 0 && (
        <div className="card">
          <strong>Remaining:</strong>{' '}
          <span>
            <em>Rp{result.remaining}</em>
          </span>
        </div>
      )}
    </Fragment>
  );
};

Display.propTypes = {
  isError: PropTypes.bool,
  result: PropTypes.shape({
    fractions: PropTypes.array.isRequired,
    remaining: PropTypes.number.isRequired,
  }),
};

Display.defaultProps = {
  isError: false,
  result: null,
};

export default Display;
