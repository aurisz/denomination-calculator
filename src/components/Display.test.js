import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import Display from './Display';

describe('Display', () => {
  const testResults = {
    totalFractions: 2,
    remaining: 0,
    fractions: [
      { count: 1, value: 10000, total: 10000 },
      { count: 1, value: 5000, total: 5000 },
    ],
  };

  it('should renders without crash', () => {
    const wrapper = shallow(<Display isError={false} result={null} />);
    expect(wrapper.exists()).toBe(true);
  });

  it('should renders if empty', () => {
    const tree = renderer
      .create(<Display isError={false} result={null} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should renders invalid input', () => {
    const tree = renderer.create(<Display isError result={null} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should renders valid results without remains', () => {
    const tree = renderer
      .create(<Display isError={false} result={testResults} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should renders valid results with remains', () => {
    testResults.remaining = 20;

    const tree = renderer
      .create(<Display isError={false} result={testResults} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
