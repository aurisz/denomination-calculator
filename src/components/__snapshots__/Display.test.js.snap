// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display should renders if empty 1`] = `
<p>
  Simple React app to calculate how many rupiah notes needed to make that amount
</p>
`;

exports[`Display should renders invalid input 1`] = `
<p>
  Invalid Input. Please try again using valid amount (e.g: 18.215, Rp17500, Rp 17.500,00)
   
</p>
`;

exports[`Display should renders valid results with remains 1`] = `
Array [
  <div
    className="card"
  >
    <span>
      1
       x 
      10000
    </span>
    <span>
      Rp
      10,000.00
    </span>
  </div>,
  <div
    className="card"
  >
    <span>
      1
       x 
      5000
    </span>
    <span>
      Rp
      5,000.00
    </span>
  </div>,
  <div
    className="card"
  >
    <strong>
      Remaining:
    </strong>
     
    <span>
      <em>
        Rp
        20
      </em>
    </span>
  </div>,
]
`;

exports[`Display should renders valid results without remains 1`] = `
Array [
  <div
    className="card"
  >
    <span>
      1
       x 
      10000
    </span>
    <span>
      Rp
      10,000.00
    </span>
  </div>,
  <div
    className="card"
  >
    <span>
      1
       x 
      5000
    </span>
    <span>
      Rp
      5,000.00
    </span>
  </div>,
]
`;
